src/hotspot/share/utilities/moveBits.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_MOVE_BITS_HPP
+#define SHARE_UTILITIES_MOVE_BITS_HPP
+
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+inline uint64_t bit_reverse(uint64_t x) {
```

Comment:
It's probably worth having comments referring to the corresponding Hacker's Delight sections for both of these functions.

Reply:
> It's probably worth having comments referring to the corresponding Hacker's Delight sections for both of these functions.

Hm, except this *isn't* the Hacker's Delight algorithm (which is the same as the seander bithacks algorithm).  Is there a reference for this algorithm?  I'd prefer something "well known" to something whose correctness I need to think hard about. I'm not spotting anything wrong here (though I think there should be more parenthesis so I don't need to think about the precedence between shift operators and bitwise-or), but @vnkozlov  is reporting failures with these changes.

Reply:
On further consideration, I don't care whether this code is correct. It should be replaced with the Hacker's Delight algorithm, which is simply better because it uses half as many large constants (which can be expensive to form on some platforms).